components/input/input-otp.component.ts

```diff
@@ -0,0 +1,228 @@
+    private readonly nzDestroyService: NzDestroyService
+  ) {
+    this.createFormGroup();
+  }
+
+  public ngOnChanges(changes: SimpleChanges): void {
+    if (changes['nzLength']?.currentValue && changes['nzLength'].currentValue !== changes['nzLength'].previousValue) {
+      this.createFormGroup();
+    }
+
+    if (changes['disabled']) {
+      this.setDisabledState(this.disabled);
+    }
+  }
+
+  onInput(index: number, event: Event): void {
+    const inputElement = event.target as HTMLInputElement;
+    const nextInput = this.otpInputs.toArray()[index + 1];
+
+    if (inputElement.value && nextInput) {
+      nextInput.nativeElement.focus();
+    } else if (!nextInput) {
+      this.selectInputBox(index);
+    }
+  }
+
+  onFocus(event: FocusEvent): void {
+    const inputElement = event.target as HTMLInputElement;
+    inputElement.select();
+  }
+
+  onKeyDown(index: number, event: KeyboardEvent): void {
+    const previousInput = this.otpInputs.toArray()[index - 1];
+
+    if (event.key === 'Backspace') {
+      event.preventDefault();
+
+      this.internalValue[index] = '';
+      this.formGroup.controls[index.toString()].setValue('', { emitEvent: false });
+
+      if (previousInput) {
+        this.selectInputBox(index - 1);
+      }
+
+      this.emitValue();
+    }
+  }
+
+  writeValue(value: string): void {
+    if (!value) return;
+
+    const controlValues = value.split('');
+    this.internalValue = controlValues;
+
+    controlValues.forEach((val, i) => {
+      const formattedValue = this.nzFormatter(val);
+      const value = this.nzMask ? this.nzMask : formattedValue;
+      this.formGroup.controls[i.toString()].setValue(value, { emitEvent: false });
+    });
+  }
+
+  registerOnChange(fn: (value: string) => void): void {
+    this.onChangeCallback = fn;
+  }
+
+  registerOnTouched(fn: () => {}): void {
+    this.onTouched = fn;
+  }
+
+  setDisabledState(isDisabled: boolean): void {
+    if (isDisabled) {
+      this.formGroup.disable();
+    } else {
+      this.formGroup.enable();
+    }
+  }
+
+  onPaste(index: number, event: ClipboardEvent): void {
+    const pastedText = event.clipboardData?.getData('text') || '';
+    if (!pastedText) return;
+
+    let currentIndex = index;
+    for (const char of pastedText.split('')) {
+      if (currentIndex < this.nzLength) {
+        const formattedChar = this.nzFormatter(char);
+        this.internalValue[currentIndex] = char;
+        const maskedValue = this.nzMask ? this.nzMask : formattedChar;
+        this.formGroup.controls[currentIndex.toString()].setValue(maskedValue, { emitEvent: false });
+        currentIndex++;
+      } else {
+        break;
+      }
+    }
+
+    event.preventDefault(); // this line is needed, otherwise the last index that is going to be selected will also be filled (in the next line).
+    this.selectInputBox(currentIndex);
+    this.emitValue();
+  }
+
+  private createFormGroup(): void {
+    this.formGroup = new FormGroup({});
```

Comment:
I think FormArray is more suitable here, WDYT?

Reply:
Yeah your right, using FormArray is the better approch.